whylog/assistant/span.py

```diff
@@ -0,0 +1,176 @@
+        :param pattern_creator: function that creates pattern for given interval and text
+        :param data_type: type of data represented by interval in text.
+        """
+        if start >= end:
+            raise SpanConstructorParamsError(start, end)
+        self.start = start
+        self.end = end
+        # length is needed many times for sorting, so counted here once.
+        self.length = end - start
+        self.is_param = is_param
+        # pattern or pattern_creator must be defined
+        if not (pattern or pattern_creator):
+            raise UnableToCreatePatternError(start, end)
+        self.pattern = pattern
+        self.pattern_creator = pattern_creator
+        self.data_type = data_type
+
+    def __repr__(self):
+        return "Span(" + str(self.start) + ", " + str(self.end) + ")"
+
+    def __eq__(self, other):
+        return self.__dict__ == other.__dict__
+
+    def __hash__(self):
+        return hash(tuple(self.__dict__.values()))
+
+    def overlaps(self, other):
+        return not (other.end <= self.start or other.start > self.start)
+
+    def update_pattern(self, text, force=False):
+        span_text = text[self.start:self.end]
+        if not self.pattern_creator:
+            return
+        if force:
+            self.pattern = self.pattern_creator(span_text)
+        else:
+            self.pattern = self.pattern or self.pattern_creator(span_text)
+
+
+def update_span_patterns(spans, text):
+    for span in spans:
+        span.update_pattern(text)
+    return spans
+
+
+def sort_by_start(spans):
+    return sorted(spans, key=attrgetter('start'))
+
+
+def sort_as_date(spans):
+    # we prefer longer dates, because it is safer (date is different in each log)
+    return sorted(spans, key=attrgetter('length'), reverse=True)
+
+
+def complementary_intervals(spans, start_index, end_index):
+    overlapping_check(spans)
+    compl_pairs = set()
+    sorted_spans = sort_by_start(spans)
+    span_start = start_index
+    for span in sorted_spans:
+        span_end = span.start
+        if span_start < span_end:
+            compl_pairs.add((span_start, span_end))
+        span_start = span.end
+    if span_start < end_index:
+        compl_pairs.add((span_start, end_index))
+    return list(compl_pairs)
+
+
+def spans_from_ranges(ranges_list, pattern_creator=None, pattern=None, is_param=False):
+    spans = [
+        Span(
+            start,
+            end,
+            pattern_creator=pattern_creator,
+            pattern=pattern,
+            is_param=is_param
+        ) for start, end in ranges_list
+    ]
+    overlapping_check(spans)
+    return list(spans)
+
+
+def overlapping_check(spans):
+    if len(spans) == 0:
+        return
+    sorted_spans = sort_by_start(spans)
+    last_span = sorted_spans[0]
+    for span in sorted_spans[1:]:
+        if last_span.overlaps(span):
+            raise OverlappingSpansError(last_span, span)
```

Comment:
what if it is not the last (first?) span that overlaps one of the other ones, but two in the middle overlap themselves? I think it needs to be n^2 or interval tree